### ISSUE TITLE
Cannot reproduce the experimental result shown in Table 1

### ISSUE THREAD

**Issue by hkiyomaru** [najoungkim/pdtb3#1]
## What

I cannot reproduce the experimental result shown in Table 1. In my environment, BERT (base, uncased) achieves 46.7% X-accuracy, which is about 3% lower than your score.

## Environment

 - OS: Ubuntu 16.04.6 LTS
 - CUDA: 10.1, V10.1.243
 - Python 3.6.5

## To reproduce

I downloaded PDTB 2.0 in CSV format and preprocessed it using your script.

```
$ bash pdtb2_setup.sh
$ python preprocess/preprocess_pdtb2.py --data_file data/pdtb2.csv --output_dir data/pdtb2_xval --split xval
row 40600
Cross-validation fold 1
Label counts:  {'Expansion.Restatement': 3206, 'Expansion.Conjunction': 3534, 'Contingency.Cause': 4172, 'Comparison.Contrast': 2120, 'Expansion.Instantiation': 1445, 'Temporal.Asynchronous': 697, 'Expansion.Alternative': 185, 'Contingency.Pragmatic cause': 83, 'Comparison.Concession': 223, 'Expansion.List': 400, 'Temporal.Synchrony': 251}
Total:  16316
...
Cross-validation fold 12
Label counts:  {'Expansion.Restatement': 3206, 'Expansion.Conjunction': 3534, 'Contingency.Cause': 4172, 'Comparison.Contrast': 2120, 'Expansion.Instantiation': 1445, 'Temporal.Asynchronous': 697, 'Expansion.Alternative': 185, 'Contingency.Pragmatic cause': 83, 'Comparison.Concession': 223, 'Expansion.List': 400, 'Temporal.Synchrony': 251}
Total:  16316
Mean train: 13676.25
Mean dev: 1280.9166666666667
Mean test: 1273.1666666666667
```

I installed `pytorch-transformers` in this repo. Before installing it, I removed line 13 of `pytorch-transformers/pytorch_transformers/__init__.py` because the line imports undeclared modules: `BertForPDTB` and `BertForPDTBClassification`.

I ran `pytorch-transformers/examples/run_pdtb.py` on each fold. I followed the parameters written in the paper.

```
$ python src/pytorch-transformers/examples/run_pdtb.py \
    --data_dir data/pdtb2_xval/fold_1 \
    --model_type bert \
    --model_name_or_path bert-base-uncased \
    --task_name pdtb2_level2 \
    --output_dir data/reproduce/pdtb2_level2/bert-base-uncased/fold_1 \
    --do_train \
    --do_eval \
    --evaluate_during_training \
    --do_lower_case \
    --per_gpu_train_batch_size 8 \
    --per_gpu_eval_batch_size 8 \
    --gradient_accumulation_steps 1 \
    --learning_rate 5e-06 \
    --weight_decay 0.0 \
    --adam_epsilon 1e-08 \
    --max_grad_norm 1.0 \
    --num_train_epochs 10.0 \
    --max_steps -1 \
    --warmup_steps 0 \
    --logging_steps 500 \
    --save_steps 500 \
    --validation_metric acc \
    --patience 5 \
    --deterministic

07/01/2020 10:49:26 - WARNING - __main__ -   Process rank: -1, device: cuda, n_gpu: 1, distributed training: False, 16-bits training: False
...
07/01/2020 11:26:55 - INFO - pytorch_transformers.modeling_utils -   loading weights file data/reproduce/pdtb2_level2/bert-base-uncased/fold_2/pytorch_model.bin
07/01/2020 11:27:03 - INFO - __main__ -   Loading features from cached file data/pdtb2_xval/fold_2/cached_dev_bert-base-uncased_128_pdtb2_level2
07/01/2020 11:27:03 - INFO - __main__ -   ***** Running evaluation  *****
07/01/2020 11:27:03 - INFO - __main__ -     Num examples = 1133
07/01/2020 11:27:03 - INFO - __main__ -     Batch size = 8
Evaluating: 100%|██████████| 142/142 [00:05<00:00, 25.06it/s]
07/01/2020 11:27:09 - INFO - __main__ -   ***** Eval results :dev *****
07/01/2020 11:27:09 - INFO - __main__ -     acc = 0.46954986760812
07/01/2020 11:27:09 - INFO - __main__ -     loss = 1.6523201620914567
07/01/2020 11:27:09 - INFO - __main__ -   Loading features from cached file data/pdtb2_xval/fold_2/cached_test_bert-base-uncased_128_pdtb2_level2
07/01/2020 11:27:09 - INFO - __main__ -   ***** Running evaluation  *****
07/01/2020 11:27:09 - INFO - __main__ -     Num examples = 1165
07/01/2020 11:27:09 - INFO - __main__ -     Batch size = 8
Evaluating: 100%|██████████| 146/146 [00:05<00:00, 24.81it/s]
07/01/2020 11:27:15 - INFO - __main__ -   ***** Eval results :test *****
07/01/2020 11:27:15 - INFO - __main__ -     acc = 0.47124463519313303
07/01/2020 11:27:15 - INFO - __main__ -     loss = 1.5675792198066842
```

I then got an X-accuracy by averaging test accuracy over folds but it was much lower than your score. (*The standard deviation shown here was calculated over folds, not different random seeds.)

```
------------  -----------------  -------------
pdtb2_level2  bert-base-uncased  46.72 (±1.31)
------------  -----------------  -------------
```

I also tried the latest version of `pytorch-transformers`, but I couldn’t reproduce the result likewise.

```
------------  -----------------  -------------
pdtb2_level2  bert-base-uncased  46.39 (±1.11)
              xlnet-base-cased   51.78 (±1.51)
------------  -----------------  -------------
```

**Comment by najoungkim** [comment]
Hi! Thanks for raising this issue. Comparing the parameters with what's in our log, the only major difference that jumps out to me is the random seed; we set seed={1, 2, 3, 4, 5} to get the results in Table 1.  With seed=1, the mean across 12 validation folds was 50.29 (std=2.02; note that the mean and std in Table 1 are from mean of means across the 5 seeds).

We did notice that even with the --deterministic flag, there were slight performance fluctuations across different machines (we believe this is an inherent GPU issue), but I remember the difference to be not as large as the results you have gotten. 

Here is the full set of parameters (with some dir names substituted) that we used for a single run of the experiment, namely cross-validation for PDTB2, Fold 2 with bert-base. For this particular run, the accuracy was 49.87.

    Training/evaluation parameters Namespace(adam_epsilon=1e-08, cache_dir='', config_name='', cuda_no=0, data_dir='my/dir', 
    device=device(type='cuda'), do_eval=True, do_lower_case=True, do_train=True, eval_all_checkpoints=False,
    evaluate_during_training=True, fp16=False, fp16_opt_level='O1', gradient_accumulation_steps=1, intermediate_weights=None,
    learning_rate=5e-06, local_rank=-1, logging_steps=500, max_grad_norm=1.0, max_seq_length=128, max_steps=-1,
    model_name_or_path='bert-base-uncased', model_type='bert', n_gpu=1, no_cuda=False, num_train_epochs=10.0,
    output_dir='data/xval_bert_base_seed1_fold_2', output_mode='classification', overwrite_cache=False,
    overwrite_output_dir=False, patience=5, per_gpu_eval_batch_size=32, per_gpu_train_batch_size=8, save_steps=500, seed=1,
    server_ip='', server_port='', task_name='pdtb2_level2', tb_logdir='', tokenizer_name='', validation_metric='acc',
    warmup_steps=0, weight_decay=0.0)

Please let us know if you still have trouble getting similar numbers! Would also be happy to look at the full log, if you send it to nkim43@jhu.edu .

**Comment by najoungkim** [comment]
Also posting the model config, in case this helps:

Model config {
"architectures": [                                                                                                                                                                                                   "BertForMaskedLM"                                                                                                                                                                                                ],                                                                                                                                                                                                                 "attention_probs_dropout_prob": 0.1,                                                                                                                                                                               "finetuning_task": "pdtb2_level2",                                                                                                                                                                                 "hidden_act": "gelu",                                                                                                                                                                                              "hidden_dropout_prob": 0.1,                                                                                                                                                                                        "hidden_size": 768,                                                                                                                                                                                                "initializer_range": 0.02,                                                                                                                                                                                         "intermediate_size": 3072,                                                                                                                                                                                         "layer_norm_eps": 1e-12,                                                                                                                                                                                           "max_position_embeddings": 512,                                                                                                                                                                                    "num_attention_heads": 12,                                                                                                                                                                                         "num_hidden_layers": 12,                                                                                                                                                                                           "num_labels": 11,                                                                                                                                                                                                  "output_attentions": false,                                                                                                                                                                                        "output_hidden_states": false,                                                                                                                                                                                     "torchscript": false,                                                                                                                                                                                              "type_vocab_size": 2,                                                                                                                                                                                              "vocab_size": 30522                                                                                                                                                                                              }

**Comment by najoungkim** [comment]
Hi Kiyomaru,

Since you've raised this issue we ran small replication experiments ourselves by setting up the repo from scratch and running the experiments on two different machines. And we got results similar in number to our originally reported result rather than the lower number you reported. 

I wonder if the issue could lie in the data rather than pytorch-transformers. I've noticed that running the preprocessing code with Python version 3.6.9 throws a UnicodeDecodeError for me. If you've had to patch this error yourself to run the preprocessing code, maybe some unicode gunk happened that made the data subtly different. Perhaps running preprocessing again with Python > 3.7.3 and rerunning the experiment (making sure to delete cached data files) would help? This is just a speculation, but something worth checking out since you were using Python 3.6.
